lib/tpm_policy.c

```diff
@@ -0,0 +1,271 @@
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+#include <errno.h>
+#include <limits.h>
+
+#include "files.h"
+#include "log.h"
+#include "tpm_hash.h"
+#include "tpm_session.h"
+#include "tpm_policy.h"
+#include "tpm2_util.h"
+
+static unsigned get_size_from_alg(TPMI_ALG_HASH hashAlg) {
+    switch (hashAlg) {
+        case TPM_ALG_SHA1:
+            return SHA1_DIGEST_SIZE;
+            break;
+        case TPM_ALG_SHA256:
+            return SHA256_DIGEST_SIZE;
+            break;
+        case TPM_ALG_SHA384:
+            return SHA384_DIGEST_SIZE;
+            break;
+        case TPM_ALG_SHA512:
+            return SHA512_DIGEST_SIZE;
+            break;
+        case TPM_ALG_SM3_256:
+            return SM3_256_DIGEST_SIZE;
+            break;
+        default:
+            LOG_ERR("Unknown hashAlg, cannot determine digest size.\n");
+            return 0;
+            break;
+    }
+}
+
+static bool evaluate_populate_pcr_digests(CREATE_POLICY_CTX *pctx, TPML_DIGEST *pcr_values) {
+    //octet value of a pcr selection group
+    uint8_t group_val=0; 
+    //total pcr indices per algorithm/ bank. Typically this is 24
+    uint8_t total_indices_for_this_alg=0; 
+    //cumulative size total of selected indices per hashAlg
+    unsigned expected_pcr_input_file_size=0;
+    //digest size returned per the hashAlg type 
+    unsigned dgst_size=0; 
+    //loop counters
+    unsigned i, j, k, dgst_cnt=0; 
+
+    //Iterating the number of pcr banks selected
+    for (i=0; i < pctx->pcr_policy_options.pcr_selections.count; i++) { 
+        //Looping to check total pcr select bits in the pcr-select-octets for a bank
+        for (j=0; j < pctx->pcr_policy_options.pcr_selections.pcrSelections[i].sizeofSelect; j++) {
+            group_val = pctx->pcr_policy_options.pcr_selections.pcrSelections[i].pcrSelect[j];
+            total_indices_for_this_alg += ( ((group_val>>3) & 1) + ((group_val>>2) & 1) +
+                ((group_val>>1) &1 ) + (group_val & 1) );
+        }
+
+        dgst_size = get_size_from_alg(pctx->pcr_policy_options.pcr_selections.pcrSelections[i].hash);
```

Comment:
unsigned dgst_size here, no need to declare it on line 73. Keep things scoped to smallest scope.